rotkehlchen/chain/ethereum/modules/curve/decoder.py

```diff
@@ -257,35 +259,155 @@ def _decode_curve_events(  # pylint: disable=no-self-use
+                event.event_subtype == HistoryEventSubType.NONE and
+                event.location_label == transaction.from_address and
+                event.counterparty in CURVE_DEPOSIT_CONTRACTS
+            ):
+                receive_event_idx = idx
+                crypto_asset = event.asset.resolve_to_crypto_asset()
+                event.event_type = HistoryEventType.RECEIVE
+                event.event_subtype = HistoryEventSubType.RECEIVE_WRAPPED
+                event.counterparty = CPT_CURVE
+                event.notes = f'Receive {event.balance.amount} {crypto_asset.symbol} after depositing in curve pool'  # noqa: E501
+                break
+
+        # exit early if there's no event for the wrapped token receipt.
+        if receive_event_idx is None:
+            log.debug(
+                f'Expected to see a receive pool token event and deposit events for a curve pool, '
+                f'but have not found them. Tx_hash: {transaction.tx_hash.hex()}',
+            )
+            return decoded_events
+
+        decoded_events[receive_event_idx].extra_data = {'deposit_events_num': len(deposit_events)}  # for accounting  # noqa: E501
+        previous_event = decoded_events[receive_event_idx]
+        for event in deposit_events:
+            maybe_reshuffle_events(previous_event, event, decoded_events)
+            previous_event = event
+
+        return decoded_events
+
+    def _handle_zap_contracts_liquidity_removal(
+            self,
+            transaction: EvmTransaction,
+            decoded_events: list[HistoryBaseEntry],
+            all_logs: list[EvmTxReceiptLog],  # pylint: disable=unused-argument
+    ) -> list[HistoryBaseEntry]:
+        """
+        This method completes the decoding of liquidity removal made through
+        Curve zap contracts such as e.g. Curve Deposit USDN, etc.
+        """
+        return_event_idx, withdrawal_events = None, []
+        for idx, event in enumerate(decoded_events):
+            if (
+                event.event_type == HistoryEventType.SPEND and
+                event.event_subtype == HistoryEventSubType.RETURN_WRAPPED and
+                event.counterparty == CPT_CURVE
+            ):
+                # exit early since transaction was decoded properly.
+                if event.extra_data is not None and 'withdrawal_events_num' in event.extra_data:
+                    return decoded_events
+
+                return_event_idx = idx
+                break
+
+        # exit early if the wrapped token event that is expected to be decoded is not found.
+        if return_event_idx is None:
+            log.debug(
+                f'Expected to see a return pool token event and withdrawal events '
+                f'for a curve pool, but have not found them. '
+                f'Tx_hash: {transaction.tx_hash.hex()}',
+            )
+            return decoded_events
+
+        # is this really needed?
+        pool_address = None
+        for tx_log in all_logs:
```

Comment:
What if a transaction has multiple remove liquidity events?

Reply:
that's the reason for my uncertainty since multiple remove liquidity events can occur